.github/workflows/cpp_extra.yml

```diff
@@ -0,0 +1,164 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+name: C++ Extra
+
+on:
+  push:
+    branches:
+      - '**'
+      - '!dependabot/**'
+    paths:
+      - '.dockerignore'
+      - '.github/workflows/cpp_extra.yml'
+      - 'ci/conda_env_*'
+      - 'ci/docker/**'
+      - 'ci/scripts/cpp_*'
+      - 'ci/scripts/install_azurite.sh'
+      - 'ci/scripts/install_gcs_testbench.sh'
+      - 'ci/scripts/install_minio.sh'
+      - 'ci/scripts/msys2_*'
+      - 'ci/scripts/util_*'
+      - 'cpp/**'
+      - 'docker-compose.yml'
+      - 'format/Flight.proto'
+      - 'testing'
+    tags:
+      - '**'
+  pull_request:
+    paths:
+      - '.dockerignore'
+      - '.github/workflows/cpp_extra.yml'
+      - 'ci/conda_env_*'
+      - 'ci/docker/**'
+      - 'ci/scripts/cpp_*'
+      - 'ci/scripts/install_azurite.sh'
+      - 'ci/scripts/install_gcs_testbench.sh'
+      - 'ci/scripts/install_minio.sh'
+      - 'ci/scripts/msys2_*'
+      - 'ci/scripts/util_*'
+      - 'cpp/**'
+      - 'docker-compose.yml'
+      - 'format/Flight.proto'
+      - 'testing'
+    types:
+      - labeled
+      - opened
+      - reopened
+      - synchronize
+  schedule:
+    - cron: |
+        0 0 * * *
+
+concurrency:
+  group: ${{ github.repository }}-${{ github.head_ref || github.sha }}-${{ github.workflow }}
+  cancel-in-progress: true
+
+permissions:
+  contents: read
+
+env:
+  ARCHERY_DEBUG: 1
+  ARROW_ENABLE_TIMING_TESTS: OFF
+  DOCKER_VOLUME_PREFIX: ".docker/"
+
+jobs:
+  check-labels:
+    name: Check labels
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    outputs:
+      ci-extra: ${{ steps.check.outputs.ci-extra }}
+    steps:
+      - name: Check
+        id: check
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          case "${GITHUB_EVENT_NAME}" in
+            push|workflow_dispatch)
+              ci_extra=true
+              ;;
+            pull_request)
+              n_ci_extra_labels=$(
+                gh pr view ${{ github.event.number }} \
+                  --jq '.labels[].name | select(. == "CI: Extra")' \
+                  --json labels \
+                  --repo ${GITHUB_REPOSITORY} | wc -l)
+              if [ "${n_ci_extra_labels}" -eq 1 ]; then
+                ci_extra=true
+              else
+                ci_extra=false
+              fi
+              ;;
+          esac
+
+          echo "ci-extra=${ci_extra}" >> "${GITHUB_OUTPUT}"
+
+  docker:
+    needs: check-labels
+    name: ${{ matrix.title }}
+    runs-on: ${{ matrix.runs-on }}
+    if: needs.check-labels.outputs.ci-extra == 'true'
```

Comment:
Instead of a separate job would it work to check for the desired label here? Something like https://stackoverflow.com/a/62331521

Reply:
You referred `types: [labeled]` approach, right?
It works only when `CI: Extra` label is added. It doesn't work when we push more commits after we add `CI: Extra` label.

https://stackoverflow.com/a/74829754 will work for the case too but we want to reuse this for nightly builds (`github.event_name == 'workflow_dispatch'`) too.
FYI: We're running `dev/tasks/tasks.yml` jobs nightly: https://lists.apache.org/list.html?builds@arrow.apache.org